ovrclk/akash
# feat: implement audited attributes

fixes #870

- test environment is based on _run/lite
- some deployment tests are busted due to change of type for Group.Requirements
- audit/handler tests TBD
- add attributes tests

working
- sign/update/delete attributes
- create deployment with attribute signature as requirement. market checks if provider can bid

limitations
- audit key built as 0x01/validator/owner address address. this causes limitation in querying list of all attributes signed by specific validator 

notes:
- proto/breakage test is expected to fail due to change of Group type

### how to test:
all examples below are tested with `_run/lite`
key `validator` is used as signer in all examples below

**prerequisites**
dump address of validator key and add it to `profiles.placement.westcoast.signedBy.anyOf[0]`
<details>
  <summary>example</summary>

  ```yaml
  ---
version: "2.0"

services:
  web:
    image: nginx
    expose:
      - port: 80
        to:
          - global: true

profiles:

  compute:
    web:
      resources:
        cpu:
          units: "100m"
        memory:
          size: "128Mi"
        storage:
          size: "256Mi"

  placement:
    westcoast:
      attributes:
        region: us-west
      signedBy:
        anyOf:
          - akash1v6aycfg99a9x4j8sltgrspmthpfzppjrn65qe9
      pricing:
        web:
          denom: uakt
          amount: 1000

deployment:
  web:
    westcoast:
      profile: web
      count: 2
  ```
  
</details>

#### sign attributes
##### values from args
```sh
# there can be as many as needed key/val pairs
../../akash --chain-id=local --keyring-backend=test --home ./cache tx audit attr create <provider_owner> key1 val1 key2 val2 --from=validator
```

#### delete attributes
##### delete all
```sh
../../akash --chain-id=local --keyring-backend=test --home ./cache tx audit delete <provider_owner> --from= validator
```

##### delete only key1
```sh
../../akash --chain-id=local --keyring-backend=test --home ./cache tx audit delete <provider_owner> key1 --from=validator
```

#### query attributes
##### all
```
../../akash --chain-id=local --home ./cache query audit list
```
